app/controllers/actions/AuthenticateGetOptionalArrivalForWriteProvider.scala

```diff
@@ -0,0 +1,88 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.actions
+
+import javax.inject.Inject
+import models.request.AuthenticatedOptionalArrivalRequest
+import models.request.AuthenticatedRequest
+import play.api.mvc._
+import play.api.mvc.Results.BadRequest
+import play.api.mvc.Results.InternalServerError
+import play.api.mvc.Results.Locked
+import repositories.ArrivalMovementRepository
+import repositories.LockRepository
+import services.ArrivalMovementService
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.xml.NodeSeq
+
+class AuthenticatedGetOptionalArrivalForWriteActionProvider @Inject()(
+  authenticate: AuthenticateActionProvider,
+  arrivalMovementRepository: ArrivalMovementRepository,
+  lockRepository: LockRepository,
+  ec: ExecutionContext
+) {
+
+  def apply(): ActionBuilder[AuthenticatedOptionalArrivalRequest, AnyContent] =
+    authenticate() andThen new AuthenticateGetOptionalArrivalForWriteAction(arrivalMovementRepository, lockRepository, ec)
+}
+
+class AuthenticateGetOptionalArrivalForWriteAction(
+  arrivalMovementRepository: ArrivalMovementRepository,
+  lockRepository: LockRepository,
+  implicit protected val executionContext: ExecutionContext
+) extends ActionFunction[AuthenticatedRequest, AuthenticatedOptionalArrivalRequest] {
+
+  override def invokeBlock[A](request: AuthenticatedRequest[A], block: AuthenticatedOptionalArrivalRequest[A] => Future[Result]): Future[Result] =
+    request.body match {
+      case body: NodeSeq =>
+        ArrivalMovementService.mrnR(body) match {
+          case None =>
+            Future.successful(BadRequest)
+
+          case Some(mrn) => {
+            arrivalMovementRepository.get(request.eoriNumber, mrn).flatMap {
+              case None => block(AuthenticatedOptionalArrivalRequest(request, None, request.eoriNumber))
+              case Some(arrival) =>
+                lockRepository.lock(arrival.arrivalId).flatMap {
+                  case false => Future.successful(Locked)
+                  case true =>
+                    block(AuthenticatedOptionalArrivalRequest(request, Some(arrival), request.eoriNumber))
+                      .flatMap {
+                        result =>
+                          lockRepository.unlock(arrival.arrivalId).map {
+                            _ =>
+                              result
+                          }
+                      }
+                      .recoverWith {
+                        case e: Exception =>
+                          lockRepository.unlock(arrival.arrivalId).map {
+                            _ =>
+                              InternalServerError
```

Comment:
Do we need a log here to indicate it couldn't unlock the record?

Reply:
The locks have a ttl of 20 seconds so they should be cleaned up in a window of 20 - 120 seconds. Given that, do you think we need that logging in our locking actions when unlocking fails? It could help us understand if we ever get those error cases where we don't unlock

Reply:
I'd stick it in otherwise you're just getting an internal server error and we're not sure what happened

Reply:
I'll follow this PR with a separate PR that does that